gapis panic

gapis panic
Target: Pixel XL, Qualcomm chipset
Reproducible : yes, when select command 317440. Not sure panic triggered by background thread or by user click action.
Gfxtrace: https://drive.google.com/open?id=0B0S0_4EgiFmcTWdvLTlJMmdKcEU


Log:
panic: runtime error: invalid memory address or nil pointer dereference
  panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x4970df1]

goroutine 180020 [running]:
github.com/google/gapid/gapis/gfxapi/gles.subApplyDynamicContextState(0x6c8e8e0, 0xc52746e900, 0x6c91880, 0xc4207aca80, 0x0, 0xc480eae8a0, 0xc4acd0ff40, 0x0, 0xc456fd3c00, 0xc420b6a030, ...)
	/Users/ggl/go/src/github.com/google/gapid/gapis/gfxapi/gles/mutate.go:5641 +0x181
github.com/google/gapid/gapis/gfxapi/gles.(*EglMakeCurrent).mutate(0xc4207aca80, 0x6c8e8e0, 0xc52746e900, 0xc480eae8a0, 0x0, 0x2fed2c7f90f72f97, 0xc4a64d1c60)
	/Users/ggl/go/src/github.com/google/gapid/gapis/gfxapi/gles/mutate.go:18016 +0x3f0
github.com/google/gapid/gapis/gfxapi/gles.(*EglMakeCurrent).Mutate(0xc4207aca80, 0x6c8e8e0, 0xc52746e900, 0xc480eae8a0, 0x0, 0x0, 0x0)
	/Users/ggl/go/src/github.com/google/gapid/gapis/gfxapi/gles/custom_replay.go:269 +0x14d
github.com/google/gapid/gapis/resolve.Memory(0x6c8e8e0, 0xc52746e900, 0xc52746e4e0, 0xc4fc49ae68, 0x4010943, 0x597bae0)
	/Users/ggl/go/src/github.com/google/gapid/gapis/resolve/memory.go:44 +0x198
github.com/google/gapid/gapis/resolve.ResolveInternal(0x6c8e8e0, 0xc52746e8a0, 0x6c8f420, 0xc52746e4e0, 0x585ee40, 0xc452376618, 0x6c8f420, 0xc52746e4e0)
	/Users/ggl/go/src/github.com/google/gapid/gapis/resolve/resolve.go:250 +0x9b0
github.com/google/gapid/gapis/resolve.ResolveService(0x6c8e8e0, 0xc52746e8a0, 0x6c8f420, 0xc52746e4e0, 0xc52746e8a0, 0x59dce60, 0x1, 0x6c64760)
	/Users/ggl/go/src/github.com/google/gapid/gapis/resolve/resolve.go:203 +0x4d
github.com/google/gapid/gapis/resolve.(*GetResolvable).Resolve(0xc452376630, 0x6c8e8e0, 0xc52746e8a0, 0x6c64760, 0xc452376630, 0x1, 0x6c692e0)
	/Users/ggl/go/src/github.com/google/gapid/gapis/resolve/get.go:35 +0xa9
github.com/google/gapid/gapis/database.(*record).resolve(0xc52746e840, 0x6c8e820, 0xc4acd0fe40, 0xc4e613fc00, 0xc52746e480)
	/Users/ggl/go/src/github.com/google/gapid/gapis/database/memory.go:72 +0x8a
github.com/google/gapid/gapis/database.(*memory).resolveLocked.func1(0xc52746e840, 0xc4acd0fe80, 0xc420371f00)
	/Users/ggl/go/src/github.com/google/gapid/gapis/database/memory.go:144 +0x45
created by github.com/google/gapid/gapis/database.(*memory).resolveLocked
	/Users/ggl/go/src/github.com/google/gapid/gapis/database/memory.go:151 +0x4ff

--- a/core/data/deep/copy.go
+++ b/core/data/deep/copy.go
@@ -21,6 +21,16 @@ import (
 	"github.com/google/gapid/core/data"
 )
 
+// Clone makes a deep copy of v.
+func Clone(v interface{}) (interface{}, error) {
+	s := reflect.ValueOf(v)
+	d := reflect.New(s.Type())
+	if err := reflectCopy(d.Elem(), s, "val", map[reflect.Value]reflect.Value{}); err != nil {
+		return nil, err
+	}
+	return d.Elem().Interface(), nil
+}
+
 // Copy recursively copies all fields, map and slice elements from the value
 // src to the pointer dst.
 func Copy(dst, src interface{}) error {

--- a/gapis/gfxapi/gles/api/extras.api
+++ b/gapis/gfxapi/gles/api/extras.api
@@ -77,15 +77,3 @@ sub void ApplyProgramInfoExtra(ProgramId programId, ref!ProgramInfo info) {
     }
   }
 }
-
-// DeviceInfo is an atom extra used to describe information about the device producing the trace.
-@internal @serialize @unused
-class DeviceInfo {
-  string ProductManufacturer
-  string ProductModel
-  string BuildDescription
-  string BuildVersionRelease
-  string BuildVersionSdk
-  string Debuggable
-  string AbiList
-}

--- a/gapis/gfxapi/gles/compat.go
+++ b/gapis/gfxapi/gles/compat.go
@@ -20,6 +20,7 @@ import (
 	"reflect"
 	"strings"
 
+	"github.com/google/gapid/core/data/deep"
 	"github.com/google/gapid/core/log"
 	"github.com/google/gapid/core/math/interval"
 	"github.com/google/gapid/core/math/u64"
@@ -271,16 +272,17 @@ func compat(ctx context.Context, device *device.Instance) (transform.Transformer
 			// TODO(dsrbecky): This might make some atoms valid for replay which were invalid on trace.
 			scs := FindStaticContextState(a.Extras())
 			if scs != nil && !version.IsES && scs.Constants.MajorVersion < 3 {
-				clone := *a
-				for i, e := range clone.extras.All() {
+				clone, err := deep.Clone(a)
+				if err != nil {
+					panic(err)
+				}
+				a = clone.(*EglMakeCurrent)
+				for _, e := range a.extras.All() {
 					if cs, ok := e.(*StaticContextState); ok {
-						scs := *cs
-						scs.Constants.MajorVersion = 3
-						scs.Constants.MinorVersion = 0
-						clone.extras[i] = &scs
+						cs.Constants.MajorVersion = 3
+						cs.Constants.MinorVersion = 0
 					}
 				}
-				a = &clone
 			}
 
 			// Mutate to set the context, Version and Extensions strings.
@@ -893,6 +895,11 @@ func compat(ctx context.Context, device *device.Instance) (transform.Transformer
 			// depending on their ability to eliminate dead code. In particular,
 			// dead code in pixel shader can allow code removal in the vertex shader.
 			// As these should have no side-effects, just drop them.
+			return
+
+		case *GlGetProgramInterfaceiv:
+			// Introduced as core in 4.3, but macOS caps out at 4.1.
+			// As this should have no side-effects, just drop them.
 			return
 
 		case *GlLabelObjectEXT,

--- a/gapis/gfxapi/gles/extras.go
+++ b/gapis/gfxapi/gles/extras.go
@@ -17,6 +17,7 @@ package gles
 import (
 	"context"
 
+	"github.com/google/gapid/core/data/deep"
 	"github.com/google/gapid/gapis/atom"
 	"github.com/google/gapid/gapis/atom/atom_pb"
 	"github.com/google/gapid/gapis/gfxapi/gles/gles_pb"
@@ -63,7 +64,11 @@ func FindErrorState(extras *atom.Extras) *ErrorState {
 func FindStaticContextState(extras *atom.Extras) *StaticContextState {
 	for _, e := range extras.All() {
 		if cs, ok := e.(*StaticContextState); ok {
-			return cs
+			clone, err := deep.Clone(cs)
+			if err != nil {
+				panic(err)
+			}
+			return clone.(*StaticContextState)
 		}
 	}
 	return nil
@@ -74,18 +79,11 @@ func FindStaticContextState(extras *atom.Extras) *StaticContextState {
 func FindDynamicContextState(extras *atom.Extras) *DynamicContextState {
 	for _, e := range extras.All() {
 		if cs, ok := e.(*DynamicContextState); ok {
-			return cs
-		}
-	}
-	return nil
-}
-
-// FindDeviceInfo searches for the DeviceInfo in the extras,
-// returning the DeviceInfo if found, otherwise nil.
-func FindDeviceInfo(extras *atom.Extras) *DeviceInfo {
-	for _, e := range extras.All() {
-		if di, ok := e.(*DeviceInfo); ok {
-			return di
+			clone, err := deep.Clone(cs)
+			if err != nil {
+				panic(err)
+			}
+			return clone.(*DynamicContextState)
 		}
 	}
 	return nil
@@ -96,7 +94,11 @@ func FindDeviceInfo(extras *atom.Extras) *DeviceInfo {
 func FindAndroidNativeBufferExtra(extras *atom.Extras) *AndroidNativeBufferExtra {
 	for _, e := range extras.All() {
 		if di, ok := e.(*AndroidNativeBufferExtra); ok {
-			return di
+			clone, err := deep.Clone(di)
+			if err != nil {
+				panic(err)
+			}
+			return clone.(*AndroidNativeBufferExtra)
 		}
 	}
 	return nil

--- a/gapis/gfxapi/gles/find_issues.go
+++ b/gapis/gfxapi/gles/find_issues.go
@@ -26,6 +26,7 @@ import (
 	"github.com/google/gapid/gapis/atom"
 	"github.com/google/gapid/gapis/atom/transform"
 	"github.com/google/gapid/gapis/capture"
+	"github.com/google/gapid/gapis/config"
 	"github.com/google/gapid/gapis/gfxapi"
 	"github.com/google/gapid/gapis/gfxapi/gles/glsl"
 	"github.com/google/gapid/gapis/gfxapi/gles/glsl/ast"
@@ -167,26 +168,28 @@ func (t *findIssues) Transform(ctx context.Context, i atom.ID, a atom.Atom, out 
 
 	switch a := a.(type) {
 	case *GlShaderSource:
-		shader := c.SharedObjects.Shaders[a.Shader]
-		var errs []error
-		var kind string
-		switch shader.Type {
-		case GLenum_GL_VERTEX_SHADER:
-			_, _, _, errs = glsl.Parse(shader.Source, ast.LangVertexShader)
-			kind = "vertex"
-		case GLenum_GL_FRAGMENT_SHADER:
-			_, _, _, errs = glsl.Parse(shader.Source, ast.LangFragmentShader)
-			kind = "fragment"
-		default:
-			t.onIssue(a, i, service.Severity_ErrorLevel, fmt.Errorf("Unknown shader type %v", shader.Type))
-		}
-		if len(errs) > 0 {
-			msgs := make([]string, len(errs))
-			for i, err := range errs {
-				msgs[i] = err.Error()
+		if !config.UseGlslang { // TODO: Verify shader with glslang
+			shader := c.SharedObjects.Shaders[a.Shader]
+			var errs []error
+			var kind string
+			switch shader.Type {
+			case GLenum_GL_VERTEX_SHADER:
+				_, _, _, errs = glsl.Parse(shader.Source, ast.LangVertexShader)
+				kind = "vertex"
+			case GLenum_GL_FRAGMENT_SHADER:
+				_, _, _, errs = glsl.Parse(shader.Source, ast.LangFragmentShader)
+				kind = "fragment"
+			default:
+				t.onIssue(a, i, service.Severity_ErrorLevel, fmt.Errorf("Unknown shader type %v", shader.Type))
 			}
-			t.onIssue(a, i, service.Severity_ErrorLevel, fmt.Errorf("Failed to parse %s shader source. Errors:\n%s\nSource:\n%s",
-				kind, strings.Join(msgs, "\n"), text.LineNumber(shader.Source)))
+			if len(errs) > 0 {
+				msgs := make([]string, len(errs))
+				for i, err := range errs {
+					msgs[i] = err.Error()
+				}
+				t.onIssue(a, i, service.Severity_ErrorLevel, fmt.Errorf("Failed to parse %s shader source. Errors:\n%s\nSource:\n%s",
+					kind, strings.Join(msgs, "\n"), text.LineNumber(shader.Source)))
+			}
 		}
 
 	case *GlCompileShader:

--- a/gapis/gfxapi/templates/api.go.tmpl
+++ b/gapis/gfxapi/templates/api.go.tmpl
@@ -80,6 +80,15 @@
   func init() {
     gfxapi.Register(API())
   }
+
+  func panicIfNotCtxCancel(err error) {
+    switch errors.Cause(err) {
+      case nil:
+      case context.Canceled:
+      default:
+        panic(err)
+    }
+  }
 {{end}}
 
 
@@ -330,6 +339,7 @@
     for i := uint64(0); i < s.count; i++ {
       {{Template "ReadStructFieldsWithRemapping" $}}
     }
+    panicIfNotCtxCancel(d.Error())
   {{end}}
 {{end}}
 
@@ -507,6 +517,7 @@
       res := make([]{{$el_ty}}, s.count)
       d := s.Decoder(ϟctx, ϟs)
       ϟmem.Read(d, &res)
+      panicIfNotCtxCancel(d.Error())
       return res
     }
 
@@ -517,6 +528,7 @@
       e := s.Slice(0, count, ϟs.MemoryLayout).Encoder(ϟs)
       ϟmem.Write(e, src[:count])
       s.OnWrite(ϟctx, ϟa, ϟs, ϟb)
+      panicIfNotCtxCancel(e.Error())
       return count
     }
 
@@ -570,9 +582,9 @@
           ϟb.StorePointer(dst, ptr)
           dst++
         }
-        if d.Error() != nil { panic(d.Error()) }
+        panicIfNotCtxCancel(d.Error())
       {{else if IsClass ($s.To | Underlying)}}
-          {{Template "ReadStructWithRemapping" $s.To}}
+        {{Template "ReadStructWithRemapping" $s.To}}
       {{else}}
         s.ReserveMemory(ϟctx, ϟa, ϟs, ϟb)
         {{if (GetAnnotation $s.To "replay_remap")}}
@@ -588,6 +600,7 @@
             ϟb.Store(ptr)
             ptr += step
           }
+          panicIfNotCtxCancel(d.Error())
         {{else}}
           ϟb.Write(s.Range(ϟl), s.ResourceID(ϟctx, ϟs))
         {{end}}
@@ -621,6 +634,7 @@
           }
           ptr += step
         }
+        panicIfNotCtxCancel(d.Error())
       {{end}}
     }
     return s

--- a/gapis/gfxapi/templates/mutate.go.tmpl
+++ b/gapis/gfxapi/templates/mutate.go.tmpl
@@ -156,7 +156,7 @@
       ϟo.ApplyReads(ϟs.Memory[ϟmem.ApplicationPool])
       {{Template "Block" $.Block}}
 
-      return nil
+      return task.StopReason(ϟctx)
     }
 
     {{if $r}}

--- a/gapis/resolve/command_tree.go
+++ b/gapis/resolve/command_tree.go
@@ -258,7 +258,9 @@ func (r *CommandTreeResolvable) Resolve(ctx context.Context) (interface{}, error
 	// Walk the list of unfiltered atoms to build the groups.
 	s := c.NewState()
 	for i, a := range c.Atoms {
-		a.Mutate(ctx, s, nil)
+		if err := a.Mutate(ctx, s, nil); err != nil && err == context.Canceled {
+			return nil, err
+		}
 		if filter(a, s) {
 			for _, g := range groupers {
 				g.process(ctx, atom.ID(i), a, s)

--- a/gapis/resolve/contexts.go
+++ b/gapis/resolve/contexts.go
@@ -79,7 +79,10 @@ func (r *ContextListResolvable) Resolve(ctx context.Context) (interface{}, error
 	s := c.NewState()
 	for i, a := range c.Atoms {
 		currentAtomIndex, currentAtom = i, a
-		a.Mutate(ctx, s, nil)
+
+		if err := a.Mutate(ctx, s, nil); err != nil && err == context.Canceled {
+			return nil, err
+		}
 
 		api := a.API()
 		if api == nil {

--- a/gapis/resolve/events.go
+++ b/gapis/resolve/events.go
@@ -39,7 +39,9 @@ func Events(ctx context.Context, p *path.Events) (*service.Events, error) {
 
 	s := c.NewState()
 	for i, a := range c.Atoms {
-		a.Mutate(ctx, s, nil)
+		if err := a.Mutate(ctx, s, nil); err != nil && err == context.Canceled {
+			return nil, err
+		}
 		// TODO: Add event generation to the API files.
 		if !filter(a, s) {
 			continue

--- a/gapis/resolve/framebuffer_changes.go
+++ b/gapis/resolve/framebuffer_changes.go
@@ -65,7 +65,9 @@ func (r *FramebufferChangesResolvable) Resolve(ctx context.Context) (interface{}
 	s := c.NewState()
 	for i, a := range c.Atoms {
 		id = atom.ID(i)
-		a.Mutate(ctx, s, nil /* no builder, just mutate */)
+		if err := a.Mutate(ctx, s, nil); err != nil && err == context.Canceled {
+			return nil, err
+		}
 		api := a.API()
 		for _, att := range allFramebufferAttachments {
 			info := framebufferAttachmentInfo{after: uint64(i)}

--- a/gapis/resolve/memory.go
+++ b/gapis/resolve/memory.go
@@ -41,7 +41,9 @@ func Memory(ctx context.Context, p *path.Memory) (*service.Memory, error) {
 
 	s := capture.NewState(ctx)
 	for _, a := range list.Atoms[:atomIdx] {
-		a.Mutate(ctx, s, nil /* no builder, just mutate */)
+		if err := a.Mutate(ctx, s, nil); err != nil && err == context.Canceled {
+			return nil, err
+		}
 	}
 
 	pool, ok := s.Memory[memory.PoolID(p.Pool)]

--- a/gapis/resolve/resource_data.go
+++ b/gapis/resolve/resource_data.go
@@ -72,7 +72,9 @@ func buildResources(ctx context.Context, p *path.Command) (*ResolvedResources, e
 	for i, a := range list.Atoms[:atomIdx+1] {
 		currentAtomResourceCount = 0
 		currentAtomIndex = uint64(i)
-		a.Mutate(ctx, state, nil /* no builder, just mutate */)
+		if err := a.Mutate(ctx, state, nil); err != nil && err == context.Canceled {
+			return nil, err
+		}
 	}
 
 	resourceData := make(map[id.ID]interface{})

--- a/gapis/resolve/state.go
+++ b/gapis/resolve/state.go
@@ -59,7 +59,9 @@ func (r *GlobalStateResolvable) Resolve(ctx context.Context) (interface{}, error
 	}
 	s := capture.NewState(ctx)
 	for _, a := range list.Atoms[:atomIdx+1] {
-		a.Mutate(ctx, s, nil)
+		if err := a.Mutate(ctx, s, nil); err != nil && err == context.Canceled {
+			return nil, err
+		}
 	}
 	return s, nil
 }
@@ -92,7 +94,9 @@ func apiState(ctx context.Context, atoms []atom.Atom, p *path.State) (interface{
 	}
 	s := capture.NewState(ctx)
 	for _, a := range atoms[:atomIdx+1] {
-		a.Mutate(ctx, s, nil)
+		if err := a.Mutate(ctx, s, nil); err != nil && err == context.Canceled {
+			return nil, err
+		}
 	}
 	res, found := s.APIs[api]
 	if !found {

